### PR TITLE
Use langchain rather than openai to allow for configuration

### PR DESCRIPTION
**Pull request by tmbo**
**Proposed changes**:
- configurable llm connection
- simple caching for llm calls to improve testing and development speed

**Status (please check what you already did)**:
- [ ] added some tests for the functionality
- [x] updated the documentation
- [ ] updated the changelog (please check [changelog](https://github.com/RasaHQ/rasa/tree/main/changelog) for instructions)
- [x] reformat files using `black` (please check [Readme](https://github.com/RasaHQ/rasa#code-style) for instructions)

### CODE REVIEW

File: rasa/utils/llm.py [hunk 1]
```diff
@@ -74,3 +80,117 @@ def sanitize_message_for_prompt(text: Optional[str]) -> str:
...
+    if "type" in custom_config:
+        # rename type to _type as "type" is the convention we use
+        # across the different components in config files.
+        # langchain expects "_type" as the key though
+        custom_config["_type"] = custom_config.pop("type")
+
+    if "_type" in custom_config and custom_config["_type"] != default_config.get(
+        "_type"
+    ):
+        return custom_config
```
Comment (twerkmeister) [inline rasa/utils/llm.py:113]: it's surprising that a different `_type` leads to all default parameters being tossed, but I guess also makes sense, many llm providers have different defaults. making a comment about this in the docs too

**Review by twerkmeister — CHANGES_REQUESTED** [review]
see comments

File: rasa/utils/llm.py [hunk 2]
```diff
@@ -74,3 +80,117 @@ def sanitize_message_for_prompt(text: Optional[str]) -> str:
...
+        The merged config.
+    """
+    if custom_config is None:
+        return default_config
+
+    if "type" in custom_config:
+        # rename type to _type as "type" is the convention we use
+        # across the different components in config files.
+        # langchain expects "_type" as the key though
+        custom_config["_type"] = custom_config.pop("type")
```
[the remark this answers is not in the document]
Comment (tmbo) [reply rasa/utils/llm.py:108]: don't want to diverge to much from the langchain format - i did make the small change to support `type` as that is what we use in other places as well, e.g.
```
nlg:
  type: rasa_plus.ml.LLMRephraser
```

**Review by twerkmeister — APPROVED** [review]
good to go, please fix the highlighted code quality issues before merge

**Pull request merged**